src/Compilers/CSharp/Portable/Binder/Binder_Constraints.cs

```diff
@@ -411,7 +411,9 @@ private TypeParameterConstraintClause GetDefaultTypeParameterConstraintClause(Ty
                 diagnostics.Add(ErrorCode.ERR_BadVisBound, location, containingSymbol, constraintType.Type);
             }
 
-            if (constraintType.Type.IsFileTypeOrUsesFileTypes() && !containingSymbol.ContainingType.IsFileTypeOrUsesFileTypes())
+            // Whether a member is effectively file-local is determined on the type level. When non-type members have constraints we need to check the containing symbol.
+            // If the 'containingSymbol.ContainingSymbol' is not a type, then 'containingSymbol' is not a member, and we don't need to enforce file types here.
+            if (constraintType.Type.IsFileTypeOrUsesFileTypes() && (containingSymbol as TypeSymbol ?? containingSymbol.ContainingSymbol as TypeSymbol)?.IsFileTypeOrUsesFileTypes() == false)
```

Comment:
>(containingSymbol as TypeSymbol ?? containingSymbol.ContainingSymbol as TypeSymbol) [](http://example.com/codeflow?start=67&length=83)

This is getting harder to read.
I'd extract this to a local function or local variable and maybe do it more formally as a switch case: for types, we check the type; for methods, we check the containing type; for any other symbol (local function), there's nothing to check.
What do you think?